### PR TITLE
Add support for retrieving credentials from `keyring`

### PR DESCRIPTION
**Pull request by BakerNet**
## Summary

Adds basic keyring auth support for `uv` commands.  Adds clone of `pip`'s `--keyring-provider subprocess` argument (using CLI `keyring` tool).

See issue: https://github.com/astral-sh/uv/issues/1520

## Test Plan

Hard to write full-suite unit tests due to reliance on `process::Command` for `keyring` cli

Manually tested end-to-end in a project with GCP artifact registry using keyring password:
```bash
➜  uv pip uninstall watchdog
Uninstalled 1 package in 46ms
 - watchdog==4.0.0

➜  cargo run -- pip install --index-url https://<redacted>/python/simple/ --extra-index-url https://<redacted>/pypi-mirror/simple/ watchdog
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/uv pip install --index-url 'https://<redacted>/python/simple/' --extra-index-url 'https://<redacted>/pypi-mirror/simple/' watchdog`
error: HTTP status client error (401 Unauthorized) for url (https://<redacted>/pypi-mirror/simple/watchdog/)

➜  cargo run -- pip install --keyring-provider subprocess --index-url https://<redacted>/python/simple/ --extra-index-url https://<redacted>/pypi-mirror/simple/ watchdog
    Finished dev [unoptimized + debuginfo] target(s) in 0.17s
     Running `target/debug/uv pip install --keyring-provider subprocess --index-url 'https://<redacted>/python/simple/' --extra-index-url 'https://<redacted>/pypi-mirror/simple/' watchdog`
Resolved 1 package in 2.34s
Installed 1 package in 27ms
 + watchdog==4.0.0
```

`requirements.txt`
```
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    .bin/generate-requirements
#
--index-url https://<redacted>/python/simple/
--extra-index-url https://<redacted>/pypi-mirror/simple/

...
```

```bash
➜  cargo run -- pip install --keyring-provider subprocess -r requirements.txt
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/uv pip install --keyring-provider subprocess -r requirements.txt`
Resolved 205 packages in 23.52s
   Built <redacted>
   ...
Downloaded 47 packages in 19.32s
Installed 195 packages in 276ms
 + <redacted>
  ...
```

### CODE REVIEW

**Comment by BakerNet** [comment]
Note:  Refactoring to use Middleware instead

**Comment by BakerNet** [comment]
Not in working state

**Review by zanieb — DISMISSED** [review]
Hi! Thank you so much for contributing and welcome to the project :)

This change overlaps with some work I've been considering, so I've left some notes about future plans — let me know how interested you are in pursuing some of the broader strokes I've outlined.

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs:349]: I implemented it more limited on purpose - because we only have access to the CLI version of keyring, the other option (python import) isn't available.  
Also, I can't say I'm a fan of pip default ("auto").  The way "auto" works in `pip`:
Try endpoint. If `401` on first request, try python import `keyring.get_credential` method.  If python import method fails, try CLI.

You could implement the keyring retry on `401` via middleware as default I guess, but... do you want to?

File: crates/uv-client/src/middleware.rs
Comment (BakerNet) [reply crates/uv-client/src/middleware.rs:52]: Makes sense, and I'd be happy to re-implement/combine the netrc middleware for this - but do you have an idea on what the interface with `uv_auth.AuthenticationStore` will look like?

One thing to be aware of here is that we do not have access to `&mut self` in the `RequestInitialiser` trait so can't call mutable methods on a struct field.  I originally planned to have store in the `KeyringMiddleware` but ran into this issue which is why it's a static ref in `get_keyring_auth`

File: crates/uv/src/main.rs
Comment (zanieb) [reply crates/uv/src/main.rs:349]: Oh I see they support `--keyring-provider subprocess` or `--keyring-provider auto` — I read this as `--keyring-provider <name>` where you could specify what command to spawn for keyring information. I'm more okay with this then.

Does `pip` use the `--keyring-provider auto` by default? i.e. does it always do keyring lookups?

Do you have more context on how the `keyring` Python library differs from the `keyring` command line tool? Happy to look into this myself some too.

File: crates/uv-client/src/middleware.rs
Comment (zanieb) [reply crates/uv-client/src/middleware.rs:52]: I hadn't figured that out yet :)

I was going to study [pip's implementation](https://github.com/pypa/pip/blob/9824a426d466680d132aa027add88ec0dda9116f/src/pip/_internal/network/auth.py). I imagine the authentication store would just need something like `get(url: Url) -> Option<Credentials>`, `set(url: Url, credentials: Credentials)`, `with_auth(url: Url) -> Url`, `save_from_url(url: Url)`?

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs:349]: > Does pip use the --keyring-provider auto by default? i.e. does it always do keyring lookups?
It does use `auto` by default - but it will only do keyring lookups if the first request fails with a `401` status code when in `auto` mode.

As for how CLI vs package differ:
The CLI tool uses the keyring python package under the hood.  The only difference is that the python package doesn't require you to supply a username to `keyring.get_credentials`,  but the CLI tool does for `keyring get.  The username passed to the CLI tool can be anything for the GCP plugin.  I don't know for sure about other cloud providers' plugins because I have not used them, but I would guess it's the same.

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs:351]: I think you're right - I don't know why it wouldn't be used for `--find-links` urls.

File: crates/uv-keyring/src/lib.rs
Comment (BakerNet) [reply crates/uv-keyring/src/lib.rs]: Done

File: crates/uv-keyring/src/lib.rs
Comment (BakerNet) [reply crates/uv-keyring/src/lib.rs]: Done

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs]: Removed `conflicts_with`

File: crates/uv-keyring/src/lib.rs
Comment (BakerNet) [reply crates/uv-keyring/src/lib.rs]: Basic move done - haven't started to tackle AuthenticationStore yet

File: crates/uv-keyring/src/lib.rs
Comment (BakerNet) [reply crates/uv-keyring/src/lib.rs]: Done

**Comment by BakerNet** [comment]
While changes are WIP, do you guys like the PR converted to Draft, or kept Open?

**Comment by zanieb** [comment]
@BakerNet if you don't want my feedback for a period of time you could mark it as a draft, otherwise I'll read through it periodically.

File: crates/uv/src/main.rs
Comment (zanieb) [reply crates/uv/src/main.rs:349]: Just wondering, why don't you like the `auto` behavior? It seems like it _could_ be sensible to avoid sending credentials to URLs that do not require it. I don't have a preference at this point but every difference from `pip` can cause confusion so I want to make an informed decision.

File: crates/uv-client/src/middleware.rs
Comment (BakerNet) [reply crates/uv-client/src/middleware.rs]: Done - implemented pretty much as you laid out, but used a slightly more verbose method name `with_url_encoded_auth`

File: crates/uv-keyring/src/lib.rs
Comment (BakerNet) [reply crates/uv-keyring/src/lib.rs]: Moved specification into type `NetLoc` which also matches `pip`'s naming (`netloc`) for their password memo dict

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs:349]: I guess it's just a personal distaste - but totally understand that differences from `pip` are problematic and can understand the point about sending credentials unnecessarily.

Implementing the retry on `401` seems non-trivial, but I'm sure it's doable.

Steps:
Create policy to retry on first `401` status (using `reqwest-retry` probably)
Before retry, delete stored auth state for URL so it will not short-circuit auth middleware
Somehow inform auth middleware to use keyring for future requests to this URL

To test:
Make sure auth middleware still runs on retry (I haven't tested this with `reqwest-retry` so don't know)

Comment (zanieb) [reply crates/uv/src/main.rs:349]: I agree that feels convoluted. I'll see if there are other opinions.

**Comment by zanieb** [comment]
Heads up that we had to change the netrc implementation to use middleware in #2325 because it caused a regression.

Let me know if you have any problems resolving the conflicts.

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs]: Changed to use same interface as `pip` - and I have a good idea of how to implement the retry, but think it would be better to do in a follow-up PR

**Comment by BakerNet** [comment]
> hey folks, happy to see this being implemented -- I'm using uv with artifact registry. Couple of questions:
> 
> 1. Do you think in future there could be an option to use a built-in credential access routine for specific hosts instead of calling a subprocess? I would really like to avoid python being involved in the process for reliability and convenience reasons. Having `keyring` tool available means there's already a python environment where it's available, meaning we would need to pin requirements for that environment because it's not just a single package but rather keyring + backends, and this creates sort of a chicken-egg problem. On top of that it's a bit of a deviation from the tool written in rust into the shaky untyped python ecosystem -- things just break randomly way more often there. Having this in uv would therefore avoid _a lot_ of dependencies and variability.
>    For example deriving a token from environment for GCP is relatively easily and such native credential helper could be much more robust, and avoid the chicken-egg thing. The only thing I'm not sure is the UX -- but I would be happy with just an arg like `--auth-helper=us-west1-python.pkg.dev=gcloud` or something like that
> 2. I haven't tried to integrate this anywhere yet but I hope the interface is flexible enough so that code that consumes uv as a dependency can plug its own auth helpers or easily reuse internal ones

Not really my place to answer, as I'm not a part of the astral team (I'm a first time contributor), but my two cents:

It seems like having to support different cloud platforms directly would be a lot of overhead to support.  I think it would make more sense to just have users figure out a way to supply their own credentials for the platform they use rather than build in GCP & Azure & AWS etc. support into `uv` itself.
I totally understand the disliking of using a python dependency to your project that needs to exist before using `uv pip install` - but adding this support helps projects migrate to `uv` more easily.  I work on a project where we are currently relying on GCP's keyring backend, and I wanted to make it easy for other people on my team to use `uv` without changing much in the rest of our environment.

**Comment by zanieb** [comment]
@vlad-ivanov-name it sounds nice to bypass `keyring` but I think we should discuss it in a new issue as long as this interface does not preclude us from doing it in the future.

**Comment by thomasgilgenast** [comment]
Just wanted to say (as only a user) thank you for the contribution and to mention that I have test-driven 177bdee and

```
UV_EXTRA_INDEX_URL=https://<location>-python.pkg.dev/<project>/<gar_name>/simple uv pip install --keyring-provider subprocess <package>
```

works perfectly with `keyrings.google_artifactregistry_auth`. I recognize that there are a lot of possible solutions to #1520 , but this one solves my use case perfectly.

The only (extremely minor) quirk I encountered is that if I include the username in the index URL

```
UV_EXTRA_INDEX_URL=https://oauth2accesstoken@<location>-python.pkg.dev/<project>/<gar_name>/simple 
```

then the command above errors with

```
error: HTTP status client error (401 Unauthorized) for url (https://<location>-python.pkg.dev/<project>/<gar_name>/simple/<package>/)
```

But when using non-uv `pip`, I do need to provide the username in `PIP_EXTRA_INDEX_URL` or else I get prompted for a username and password

```
User for <location>-python.pkg.dev: 
```

But to be clear this is not a problem for me at all, just something that could be a useful piece of info for users trying to convert from pip to uv who stumble upon this PR. (If anything it was inconvenient to me that pip required the username - maybe it is actually a bug specific to `keyrings.google_artifactregistry_auth` when combined with `--keyring-provider subprocess`?)

And possibly totally unrelated to this PR, if there might be a `UV_KEYRING_PROVIDER` env var parallel to `PIP_KEYRING_PROVIDER`, in analogy to `UV_EXTRA_INDEX_URL`::`PIP_EXTRA_INDEX_URL` that would be very convenient for me. But it may be something for a separate PR or to discuss before implementing (cost of additional env vars versus ability to consistently control `uv` behavior with env vars). I have opened a minimal PR https://github.com/BakerNet/uv/pull/1 against your branch @BakerNet just for my own experimentation. Please feel free to close if you feel that I'm sneaking something in there that doesn't belong with this PR or deserves a separate discussion.

File: crates/uv-auth/src/lib.rs
Comment (zanieb) [reply crates/uv-auth/src/lib.rs:163]: I guess they make less sense with the `NetLoc` implementation?

**Comment by BakerNet** [comment]
> I have opened a minimal PR [BakerNet#1](https://github.com/BakerNet/uv/pull/1) against your branch @BakerNet just for my own experimentation. Please feel free to close if you feel that I'm sneaking something in there that doesn't belong with this PR or deserves a separate discussion.

Seems like a totally uncontroversial QoP improvement to me, so I merged it in.  Thanks for testing this out, @thomasgilgenast  btw!

File: crates/uv-auth/src/store.rs
Comment (zanieb) [reply crates/uv-auth/src/store.rs:13]: Why do we store `None` credentials in here?

File: crates/uv/src/main.rs
Comment (zanieb) [inline crates/uv/src/main.rs:367]: I'm a little hesitant on this interface. We might want to consider what our ideal interface for this then just include a dedicated interoperability flag for pip's interface.

I think in part this depends on our future plans around keyring support (which I know there was some discussion about already). In particular, if we use something like the Rust `keyring` crate do we need to provide any options other than a boolean toggle? Perhaps we do still want to allow the user to select a specific provider?

File: crates/uv/src/main.rs
Comment (zanieb) [reply crates/uv/src/main.rs]: I'd be curious to hear @charliermarsh's opinion regarding only searching for authentication on 401s.

It does seem brittle to me, we've seen all sorts of quirky behavior from package indexes when unauthorized.

File: crates/uv-auth/src/lib.rs
Comment (BakerNet) [reply crates/uv-auth/src/lib.rs:163]: Yeah, because it's now an equality comparison, order doesn't matter.  One of the `NetLoc` is the trusted url, one is the new url, but which side of the equality operation it's on is logically irrelevant.

File: crates/uv-auth/src/store.rs
Comment (BakerNet) [reply crates/uv-auth/src/store.rs:13]: > Why do we store `None` credentials in here?

Just so we don't re-check netrc or keyring when they already failed

Comment (zanieb) [reply crates/uv-auth/src/store.rs:13]: That makes sense, thanks!

File: crates/uv-auth/src/keyring.rs
Comment (BakerNet) [reply crates/uv-auth/src/keyring.rs:22]: > Do you think that it's worth using something like [docs.rs/keyring/latest/keyring](https://docs.rs/keyring/latest/keyring/) instead of a subprocess?

Instead?  No.  In addition?  Yes.

As mentioned in issue https://github.com/astral-sh/uv/issues/1520 - the cloud providers provide backend plugins for the python keyring tool.
Using something like https://docs.rs/keyring/latest/keyring/ would be fine for the `Import` case, but supporting a command-line `keyring` should still exist regardless (especially if you're trying to keep parity with pip) and if someone makes a CLI keyring with cloud provider support based on `keyring-rs` in the future, cool.  If it has the same CLI interface, it should just work.  If it has a different name, add to the `KeyringProvider`s or implement the custom version like discussed with @vlad-ivanov-name above.

File: crates/uv/src/main.rs
Comment (BakerNet) [reply crates/uv/src/main.rs:367]: I feel like a value enum is a pretty flexible interface - you can just add additional options to the enum and change logic based on the variant.  If there are additional options, you can just add another argument that relies on a specific `KeyringProvider` value like `--keyring-provider custom --custom-keyring-command keyring-rs` as a possible example (I'm certain `clap` supports this)

File: crates/uv-auth/src/keyring.rs
Comment (BakerNet) [reply crates/uv-auth/src/keyring.rs:70]: Fair point - that would be more accurate

File: crates/uv-auth/src/keyring.rs
Comment (zanieb) [reply crates/uv-auth/src/keyring.rs:22]: 👍 makes sense

File: crates/uv-auth/src/keyring.rs
Comment (BakerNet) [reply crates/uv-auth/src/keyring.rs]: Done

File: crates/uv-auth/src/middleware.rs
Comment (zanieb) [inline crates/uv-auth/src/middleware.rs]: If you do the refactor to separate `None` from the `Error` case we should be able to log failures here.

Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs]: What is the correct way to log errors in `uv` codebase, btw?  `warn!` or `warn_user!` maybe?

Comment (zanieb) [reply crates/uv-auth/src/middleware.rs]: Hm that's tough.

Ideally we'd attach this as a hint to request failures so when we _error_ we'd be able to suggest why. In the meantime, I think a tracing message is the way to go unless it's very actionable for the user to fix the problem. I believe the difference is a tracing message will only show up with `-v` but `warn_user!` will always display.

**Comment by zanieb** [comment]
Thanks for addressing all the feedback so promptly.

I've added @charliermarsh as a reviewer — once he's happy we'll be good to go.

File: crates/uv-auth/src/middleware.rs
Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs]: Done

File: crates/uv/src/main.rs
Comment (charliermarsh) [inline crates/uv/src/main.rs:1247]: We should add this to the list of supported environment variables in the README.

File: crates/uv-auth/src/middleware.rs
Comment (charliermarsh) [inline crates/uv-auth/src/middleware.rs:57]: For my own clarification... this will propagate the credentials based purely on `host` etc., right? In other words, the URLs do not need to be identical, is that correct? Just the same "realm" and so on?

Is this generally a safe and accepted thing to do? (That is: propagate the credentials here.) It's a little different than what we do on `main` which is more limited: when we make a request, we propagate credentials to the URLs that are found on the page returned by the request (like `with_url_encoded_auth`, I think).

File: crates/uv-auth/src/store.rs
Comment (charliermarsh) [inline crates/uv-auth/src/store.rs:56]: When can these fail? (These will be effectively silent for the user.)

File: crates/uv-auth/src/middleware.rs
Comment (charliermarsh) [inline crates/uv-auth/src/middleware.rs:67]: Can you expand on the comment here? Should this be `unreachable!`?

File: crates/uv-auth/src/middleware.rs
Comment (charliermarsh) [inline crates/uv-auth/src/middleware.rs:162]: Should these unwraps just be `?` given that we already return `Result` here?

File: crates/uv-auth/src/store.rs
Comment (charliermarsh) [inline crates/uv-auth/src/store.rs:32]: I think it's a bit inflexible for these to clone. Callers can always clone if they need owned data, but callers that _don't_ need owned data are now _forced_ to clone. I think these should return references.

File: crates/uv-auth/src/store.rs
Comment (charliermarsh) [inline crates/uv-auth/src/store.rs:43]: This is kind of a hidden allocation, because we're taking a reference and then cloning inside the `From`. I think it would be clearer as `From<Authenticator>`, personally.

File: crates/uv-auth/src/store.rs
Comment (charliermarsh) [inline crates/uv-auth/src/store.rs:71]: I'm not a huge fan of the singleton struct here but I don't need to block the PR on it. It's just not something we've really done anywhere else that I can remember. Maybe @BurntSushi would have a better opinion on it.

File: crates/uv-auth/src/middleware.rs
Comment (charliermarsh) [inline crates/uv-auth/src/middleware.rs:92]: Nit: can you inline these please? `debug!("No keyring credentials found for: {url}");`

File: crates/uv/src/main.rs
Comment (charliermarsh) [reply crates/uv/src/main.rs]: I haven't thought about it deeply but I have a negative reaction to only sending credentials after seeing a 401.

File: crates/distribution-types/src/file.rs
Comment (charliermarsh) [inline crates/distribution-types/src/file.rs:58]: So the assumption here is that we've already hit the base URL on a previous invocation of the middleware, and so its credentials are already stored in the `AuthenticationStore`, and will be propagated here if needed?

**Review by charliermarsh — COMMENTED** [review]
Thanks for this! I think I need to better understand some of the credential flows before merging.

File: crates/uv-auth/src/middleware.rs
Comment (zanieb) [reply crates/uv-auth/src/middleware.rs:57]: Yeah we're using the "net loc" which is the same "realm". This is compliant with the relevant web standard RFC and AFAIK exactly what pip does. It seems totally fine to me but I appreciate the extra scrutiny.

File: crates/uv-auth/src/store.rs
Comment (zanieb) [reply crates/uv-auth/src/store.rs:71]: I imagined the passwords would live on it but I think that's infeasible for some reason?

File: crates/uv-auth/src/middleware.rs
Comment (charliermarsh) [reply crates/uv-auth/src/middleware.rs:57]: Do you know why we _also_ need to call `with_url_encoded_auth` before the middleware then?

File: crates/uv/src/main.rs
Comment (zanieb) [reply crates/uv/src/main.rs:1247]: We'll also want to document support and precedence in a registry authentication section following https://github.com/astral-sh/uv?tab=readme-ov-file#git-authentication but I can do that in a follow-up (just posting so I don't forget).

File: crates/uv-auth/src/middleware.rs
Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs:57]: I didn't follow the full flow after the `url` is copied - I saw it was being used in `File` structs and didn't want to potentially introduce regression by assuming theses `urls` having the auth copied to were going to be used in the same run of `uv`

File: crates/uv-auth/src/middleware.rs
Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs:67]: The idea here is, if the url does not come to the middleware with URL-encoded auth, we shouldn't introduce URL-encoded auth because it's an inherently unsafe way to send auth data.  On main, there was already code copying the auth from trusted urls onto response/body urls so this line shouldn't be hit under normal cicumstances anyways.

File: crates/uv-auth/src/store.rs
Comment (BakerNet) [reply crates/uv-auth/src/store.rs:56]: I didn't look into this - it's just copied over from the main (lib.rs) function

Comment (charliermarsh) [reply crates/uv-auth/src/store.rs:56]: Okay thanks, that's a perfectly fine answer :)

File: crates/uv-auth/src/store.rs
Comment (BakerNet) [reply crates/uv-auth/src/store.rs:71]: My initial plan was for a store instance to be a field of the `AuthMiddleware` struct and the store to hold an instance of the password table - but you do not have access to mutable self in the `Middleware` trait, so that was the limitation I ran into.  Combine that with keeping the copying from trusted URL that wouldn't have access to the same instance - I went with this approach.

File: crates/distribution-types/src/file.rs
Comment (BakerNet) [reply crates/distribution-types/src/file.rs:58]: Yes, that was my assumption.  That if `base` has auth, it received it from a copy from a previous request (in `FlatIndexClient.read_from_url` or `RegistryClient.simple_single_index`).

File: crates/uv-auth/src/middleware.rs
Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs]: Done

File: crates/uv-auth/src/store.rs
Comment (BakerNet) [reply crates/uv-auth/src/store.rs:43]: Done

File: crates/uv-auth/src/store.rs
Comment (BakerNet) [reply crates/uv-auth/src/store.rs:32]: Done

File: crates/uv-auth/src/middleware.rs
Comment (BakerNet) [reply crates/uv-auth/src/middleware.rs]: Done

File: crates/uv-auth/src/store.rs
Comment (zanieb) [reply crates/uv-auth/src/store.rs:71]: Can you just use interior mutability to get around this? e.g. `RefCell<T>`?

Comment (charliermarsh) [reply crates/uv-auth/src/store.rs:71]: I believe you could just do it with the Mutex within the struct. But we can change this later if desired.

**Review by charliermarsh — APPROVED** [review]
I'm cool with this, thanks for the nice PR! I defer to @zanieb to merge.

**Pull request merged**
